src/uniqlist.c

```diff
@@ -200,6 +200,11 @@ SEXP rleid(SEXP l, SEXP cols) {
           // 8 bytes of bits are identical. For real (no rounding currently) and integer64
           // long long == 8 bytes checked in init.c
           break;
+        case CPLXSXP: {
+          // tried to make long long complex * but got a warning that it's a GNU extension
+          Rcomplex *pz = COMPLEX(jcol);
+          same = (long long)pz[i].r == (long long)pz[i-1].r && (long long)pz[i].i == (long long)pz[i-1].i;
```

Comment:
This `(long long)` is a _cast_ and won't achieve the desired op.  This cast will trunc 3.4 to 3 for example.  What's needed is a type _pun_, which is what the REALSXP case just above does.

Reply:
So to clarify in the `REALSXP` branch:

```
long long *ll = (long long *)REAL(jcol);
same = ll[i]==ll[i-1];
```

What's happening here is `REAL` is _actually_ a pointer to `double`, but `ll` is a `long long` pointer to the same address as where `jcol` is.

So then `ll[i] == ll[i-1]` treats the elements of `jcol` (which are actually real) as 8-byte integers & compares those 64 bits.

Is that the gist of it?

Reply:
Yes: exactly.

Reply:
Is there a native 16-byte type to pun to? I'm not seeing one. I guess we could just use pure equality-on-doubles & leave it at that as it's low enough priority.

Reply:
Another way, since all the cases have the same code other than the comparison, is to take away all the cases and just do one `memcmp` for all types (just pass the size to memcmp).  Would need benchmarking on large data to see if it was any slow.   (I didn't look closely to see if there are any other difference between all those cases in this switch in this file.)